Core/Object Arts/Dolphin/DolphinSure/BCryptLibrary.cls

```diff
@@ -0,0 +1,183 @@
+
+	NTSTATUS WINAPI BCryptCreateHash(
+		_Inout_	BCRYPT_ALG_HANDLE	hAlgorithm,
+		_Out_	BCRYPT_HASH_HANDLE*	phHash,
+		_Out_	PUCHAR				pbHashObject,
+		_In_opt_	ULONG				cbHashObject,
+		_In_opt_	PUCHAR				pbSecret,
+		_In_		ULONG				cbSecret,
+		_In_		ULONG				dwFlags
+);"
+	<cdecl: dword BCryptCreateHash handle handle* byte* dword byte* dword dword>
+	^ self invalidCall.!
+
+destroyHash: aHashHandle
+	"The BCryptDestroyHash function closes an algorithm provider.
+
+	NTSTATUS WINAPI BCryptDestroyHash(
+		_Inout_	BCRYPT_HASH_HANDLE	hHash );"
+
+	<cdecl: dword BCryptDestroyHash handle>
+	^ self invalidCall.!
+
+finishHash: aHashHandle output: aOutput outputSize: aOutputSize flags: aFlags
+	"The BCryptFinishHash function retrieves the hash or Message Authentication Code (MAC) value for the data accumulated from prior calls to BCryptHashData.
+
+	NTSTATUS WINAPI BCryptFinishHash(
+		_Inout_	BCRYPT_HASH_HANDLE	hHash,
+		_Out_	PUCHAR				pbOutput,
+		_In_		ULONG				cbOutput,
+		_In_		ULONG				dwFlags );"
+
+	<cdecl: dword BCryptFinishHash handle byte* dword dword>
+	^ self invalidCall.!
+
+getProperty: aHandle property: aProperty output: aOutput outputSize: aOutputSize resultSize: aResultSize flags: aFlags
+	"The BCryptGetProperty function retrieves the value of a named property for a BCrypt object.
+
+	NTSTATUS WINAPI BCryptGetProperty(
+		_In_		BCRYPT_HANDLE	hObject,
+		_In_		LPCWSTR			pszProperty,
+		_Out_	PUCHAR			pbOutput,
+		_In_		ULONG			cbOutput,
+		_Out_	ULONG			*pcbResult,
+		_In_		ULONG			dwFlags ); "
+
+	<cdecl: dword BCryptOpenAlgorithmProvider handle bstr byte* dword dword* dword>
+	^ self invalidCall.!
+
+hashData: aHashHandle input: aInput inputSize: aInputSize flags: aFlags
+	"The BCryptHashData function performs a one way hash or Message Authentication Code (MAC) on a data buffer.
+
+	NTSTATUS WINAPI BCryptHashData(
+		_Inout_	BCRYPT_HASH_HANDLE	hHash,
+		_In_		PUCHAR				pbInput,
+		_In_		ULONG				cbInput,
+		_In_		ULONG				dwFlags );"
+
+	<cdecl: dword BCryptHashData handle byte* dword dword>
+	^ self invalidCall.!
+
+openAlgorithmProvider: aHandle id: anId implementation: anImplementation flags: aFlags
+	"The BCryptOpenAlgorithmProvider function loads and initializes a BCrypt provider.
+
+	NTSTATUS WINAPI BCryptOpenAlgorithmProvider(
+		_Out_	BCRYPT_ALG_HANDLE*	phAlgorithm,
+		_In_		LPCWSTR				pszAlgId,
+		_In_		LPCWSTR				pszImplementation,
+		_In_		DWORD				dwFlags );"
+
+	<cdecl: dword BCryptOpenAlgorithmProvider handle* bstr bstr dword>
+	^ self invalidCall.!
+
+shaAlgorithmHandle
+	"Answer internal handle. to the SHA1 algorithm.
+	Create it, if it was not allocated yet.
+	This handle is never freed, because re-creating it is expensive."
+
+	| status |
+
+	( shaAlgorithmHandle isNil or: [ shaAlgorithmHandle value = 0 ] ) ifTrue: [
+		shaAlgorithmHandle := ExternalHandle new.
+		status := self openAlgorithmProvider: shaAlgorithmHandle id: BCRYPT_SHA1_ALGORITHM implementation: nil flags: 0.
```

Comment:
I realise this is for compatibility, but we should stop using SHA1. I don't think we should be storing this algorithm handle in the library either. We generally don't put any state or functionality (other than the FFI calls) in the library class. It would make more sense to store the algorithm handle in the SecureHashAlgorithm class, and then perhaps have support for different algorithms too. I think this is Ok for now though (it is an improvement as is).